// knip@6.6.0/dist/util/glob-cache.js
import { createHash } from 'node:crypto';
import fs from 'node:fs';
import path from 'node:path';
import { deserialize, serialize } from 'node:v8';
import { version } from '../version.js';
import { debugLog } from './debug.js';
import { isDirectory, isFile } from './fs.js';
import { dirname } from './path.js';
const CACHE_FILENAME = `glob-${version}.cache`;
let cacheFilePath;
let cache;
let isDirty = false;
export const initGlobCache = (cacheLocation) => {
    cacheFilePath = path.resolve(cacheLocation, CACHE_FILENAME);
    if (isFile(cacheFilePath)) {
        try {
            cache = deserialize(fs.readFileSync(cacheFilePath));
        }
        catch {
            debugLog('*', `Error reading glob cache from ${cacheFilePath}`);
            cache = new Map();
        }
    }
    else {
        cache = new Map();
    }
};
export const isGlobCacheEnabled = () => cache !== undefined;
export const computeGlobCacheKey = (input) => {
    const h = createHash('sha1');
    h.update(input.cwd);
    h.update('\0');
    h.update(input.dir);
    h.update('\0');
    h.update(input.gitignore ? '1' : '0');
    h.update('\0');
    for (const p of input.patterns) {
        h.update(p);
        h.update('\0');
    }
    return h.digest('base64url');
};
const validateEntry = (entry) => {
    for (const dir in entry.dirMtimes) {
        try {
            const stat = fs.statSync(dir);
            if (stat.mtimeMs !== entry.dirMtimes[dir])
                return false;
        }
        catch {
            return false;
        }
    }
    return true;
};
export const getCachedGlob = (key) => {
    if (!cache)
        return undefined;
    const entry = cache.get(key);
    if (!entry)
        return undefined;
    if (!validateEntry(entry)) {
        cache.delete(key);
        isDirty = true;
        return undefined;
    }
    return entry.paths;
};
const captureDirMtimes = (paths, baseDir) => {
    const dirs = new Set();
    dirs.add(baseDir);
    for (const p of paths) {
        let d = dirname(p);
        while (d.length >= baseDir.length) {
            if (dirs.has(d))
                break;
            dirs.add(d);
            const parent = dirname(d);
            if (parent === d)
                break;
            d = parent;
        }
    }
    const result = {};
    for (const d of dirs) {
        try {
            const stat = fs.statSync(d);
            if (stat.isDirectory())
                result[d] = stat.mtimeMs;
        }
        catch {
        }
    }
    return result;
};
export const setCachedGlob = (key, paths, baseDir) => {
    if (!cache)
        return;
    cache.set(key, {
        paths,
        dirMtimes: captureDirMtimes(paths, baseDir),
    });
    isDirty = true;
};
export const clearGlobCache = () => {
    if (cache) {
        cache.clear();
        isDirty = true;
    }
};
export const flushGlobCache = () => {
    if (!cache || !cacheFilePath || !isDirty)
        return;
    try {
        const dir = dirname(cacheFilePath);
        if (!isDirectory(dir))
            fs.mkdirSync(dir, { recursive: true });
        fs.writeFileSync(cacheFilePath, serialize(cache));
        isDirty = false;
    }
    catch {
        debugLog('*', `Error writing glob cache to ${cacheFilePath}`);
    }
};
